internal/provider/resource_storagezone.go

```diff
@@ -0,0 +1,399 @@
+	return fmt.Errorf(message, key, old, new, key)
+}
+
+func immutableReplicationRegionError(key string, removed []interface{}) error {
+	const message = "'%s' can be added but not removed once the zone has been created.\n" +
+		"This error occurred when attempting to remove values %+q from '%s'.\n" +
+		"To remove an existing '%s' the 'bunny_storagezone' must be deleted and recreated.\n" +
+		"WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!"
+	return fmt.Errorf(
+		message,
+		key,
+		removed,
+		key,
+		key,
+	)
+}
+
+func resourceStorageZoneCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	clt := meta.(*bunny.Client)
+
+	originURL := getStrPtr(d, keyOriginURL)
+	if !d.HasChange(keyOriginURL) {
+		originURL = nil
+	}
+
+	sz, err := clt.StorageZone.Add(ctx, &bunny.StorageZoneAddOptions{
+		Name:               getStrPtr(d, keyName),
+		OriginURL:          originURL,
+		Region:             getStrPtr(d, keyRegion),
+		ReplicationRegions: getStrSetAsSlice(d, keyReplicationRegions),
+	})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("creating storage zone failed: %w", err))
+	}
+
+	d.SetId(strconv.FormatInt(*sz.ID, 10))
+
+	// StorageZone.Add() only supports to set a subset of a Storage Zone object,
+	// call Update to set the remaining ones.
+	if diags := resourceStorageZoneUpdate(ctx, d, meta); diags.HasError() {
+		// if updating fails the sz was still created, initialize with the SZ
+		// returned from the Add operation
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "setting storage zone attributes via update failed",
+		})
+
+		if err := storageZoneToResource(sz, d); err != nil {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "converting api-type to resource data failed: " + err.Error(),
+			})
+		}
+
+		return diags
+	}
+
+	return nil
+}
+
+func resourceStorageZoneUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	clt := meta.(*bunny.Client)
+
+	storageZone := storageZoneFromResource(d)
+
+	id, err := getIDAsInt64(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	updateErr := clt.StorageZone.Update(ctx, id, storageZone)
+	if updateErr != nil {
+		// if our update failed then revert our values to their original
+		// state so that we can run an apply again.
+		revertErr := revertUpdateValues(d)
+
+		if revertErr != nil {
+			return diagsErrFromErr("updating storage zone via API failed", revertErr)
+		}
+
+		return diagsErrFromErr("updating storage zone via API failed", updateErr)
+	}
+
+	updatedStorageZone, err := clt.StorageZone.Get(ctx, id)
+	if err != nil {
+		return diagsErrFromErr("fetching updated storage zone via API failed", err)
+	}
+
+	if err := storageZoneToResource(updatedStorageZone, d); err != nil {
+		return diagsErrFromErr("converting api type to resource data after successful update failed", err)
+	}
```

Comment:
So I did that because looking at the pull zone as a template that update returns a response containing the pz, which is then fed back into the state update function. Because SZ update returns a 204 I assumed I would need to read the SZ after update to write it. But in retrospect I think you're right and that TF already knows the new state.

Reply:
@fho I pushed a commit that removes this block.